verify.js

```diff
@@ -86,92 +51,134 @@ module.exports = function (jwtString, secretOrPublicKey, options, callback) {
 
   var header = decodedToken.header;
 
-  if (!~options.algorithms.indexOf(header.alg)) {
-    return done(new JsonWebTokenError('invalid algorithm'));
+  if(header.alg ==='none' && decodedToken.signature) {
```

Comment:
Let's try to do PRs as atomic as possible. This change is not related, if you are interested on it, please open a different PR.

Reply:
I believe the change is related. Because the callback could unintentionally return undefined (for example it should return an error but instead it returns undefined) then suddenly all unsigned keys are accepted.

Reply:
I see, but we should use all checks we currently use for that https://github.com/auth0/node-jsonwebtoken/blob/master/verify.js#L52-L91 Otherwise it would mean there is a bug in the current code.

Reply:
Sorry I don't know what you mean to say. I don't think there is a bug in the current code but I do think the logic was off. And adding `none` to the algorithms within the `decode` function scares me. That isn't a problem in the current code, but with this particular PR it can easily compromise security.

One has to be very very strict about when to allow unsigned JWT's and consider any way how they could sneak in unintentionally. I've thought long about this and looked up the JWT / JWS standards and I think this is the proper way to do it. 

We could discuss about if this should be in a separate PR or not, but let's not be too bureaucratic about it. If you agree with the code than why make a fuzz. We all have busy schedules, you know.

Reply:
Let's take another approach to see if I explain better my thoughts:
What's the risk difference between passing a secret/key nowadays vs passing the secret/key using the new function?

Reply:
Ah, I thought you already got that 😉. It's actually a security hole, not a risk. I made it more explicit now.

Before, if you passed an unsigned token AND a secret, it would be blocked. If you pass an unsigned token and no secret it would always pass.

Now with the callback code, when you pass an unsigned token, there will not be a secret and the token is valid. So to block it, it checks if unsigned tokens are allowed.

💡 or 🤔?

Reply:
Better, now we have the rationale of the change.
> Now with the callback code, when you pass an unsigned token, there will not be a secret

Why won't there be a secret? I can see in these use cases, let me know if you have others in mind:
(For all of them the input is an unsigned token with algorithm `none`)
- Your verification requires a signed token, `getSecret` returns a secret => unsigned token + secret => verification fails.
- Your verification requires a signed token, `getSecret` can't find a secret to use => error from `getSecret` => verification fails.
- Your verification allows unsigned tokens => You don't pass `getSecret` function, if you do the function needs to return nothing as secret.

In your last commit in case of algorithm === `none` then `getSecret` returns undefined, in that case the door is just open to anyone to craft a JWT with same data as an expected signed JWT and `none` as algorithm, and since `getSecret` from consumer won't be called the token will pass.

I have a proposal for you, clean all the checks you have added or modified, leave just the call to `getSecret` and move the current checks where they should be in the new code. Starting from that point try to write tests which will make the token end up as "verified" when it shouldn't, or the other way around, then we can analyze them and build the fixes properly.

Reply:
Sorry for the delay, I was kind of MIA.

Point 1: the `getSecret` function would normally use `kid` to determine which secret to return, but since algorithm is `none` and `kid` is absent, there's no secret to return. So then it would have to return a bogus key which I think doesn't make much sense.

Point 2: Theoretically it is a possibility but I think it is the wrong approach. The `getSecret` function should have a single, clear and intuitive purpose: getting the secret. It should therefor return a secret, and if that is not possible it should return an error. Nothing else. What you are proposing here is to give `getSecret` a second purpose, namely, to determine if an unsigned token is allowed or not. It is not intuitive and especially prone to disasters. For instance `return keys[kid]` is all it takes to leave the door wide open.

Point 3: By not supplying the `getSecret` function it is possible to allow ONLY unsigned tokens. In that case it would support only unsigned tokens, or only signed tokens but not both. If you want so support both by returning nothing (I think you mean 'undefined') then I think it's a bad idea, see point 2.

So I think there are two possibilities. 

1. Allow only unsigned tokens OR only signed tokens, but not both. If the `getSecret` function is present, it will allow only signed tokens. If it is `undefined` it would allow only unsigned tokens, just like in the first part of your 3rd point.

2. Allow both unsigned and signed tokens at the same time. Allow unsigned tokens only based on a configuration setting, not based on the result of `getSecret`. Like in my PR, unsigned tokens are only allowed if the `algorithms` array contains `none`.

I think unsigned tokens only have use for testing and either way makes that possible but I think method 1 would be the least error prone.

Reply:
From my previous comment:
> In your last commit in case of algorithm === none then getSecret returns undefined, in that case the door is just open to anyone to craft a JWT with same data as an expected signed JWT and none as algorithm, and since getSecret from consumer won't be called the token will pass.

Re-reading all the code I've seen it would fail with `invalid algorithm: none`, therefore forget about that comment then.

We would need a major release for all these changes, since:
- Tokens with `none` as algorithm will be accepted only when it is passed through `options.algorithms`.
- Changes on errors.

I've seen a couple of things to mention. I'll do it on their lines today.